### ISSUE TITLE
[feat] support Azure Trusted Signing for signing tauri app

### ISSUE THREAD

**Issue by anatawa12** [tauri-apps/tauri#9578]
### Describe the problem

Many certificate authority sells certificates with high price.
Recently Microsoft released Azure Trusted Signing for code signing and it's relatively cheap.

However, since signing tauri apps are completely integrated inside tauri cli and tauri cli doesn't have option to customize `/dlib`, `/dmdf` parameters for signtool, it looks we cannot use Azure Trusted Signing.

### Describe the solution you'd like

Add some option to sign with Azure Trusted Signing

### Alternatives considered

I think it might also be good to have some option to customize command-line options for signtool or use complately custom command for code signing.

### Additional context

https://techcommunity.microsoft.com/t5/security-compliance-and-identity/trusted-signing-is-in-public-preview/ba-p/4103457
https://learn.microsoft.com/en-us/azure/trusted-signing/how-to-signing-integrations

**Comment by FabianLars** [comment]
This PR will be converted into something generic before being merged https://github.com/tauri-apps/tauri/pull/8718

At least once MS lifts the stupid 3-years tax history requirement, this should have explicit support in tauri imo (so that it's a direct config option that doesn't require the use of the generic option)

**Comment by Levminer** [comment]
As a workaround I created a [CLI](https://github.com/Levminer/trusted-signing-cli). It requires the azure cli, but downloads the required dlib for the Trusted Signing. Which gets called in the tauri.conf.json:
```json
{
	"build": {
		"beforeBundleCommand": "trusted-signing-cli -e <url> -a <account name> -c <certificate profile name> path/to/exe/file1.exe",
	},
```

cc @FabianLars: I could integrate this into Tauri, should I create a PR?

**Comment by FabianLars** [comment]
I don't think we'll add any specific solutions into tauri now with the generic solution (not sure). What i would love though is to collect different examples for the generic command (from the aforementioned #9865) in the docs.

**Comment by Levminer** [comment]
> I don't think we'll add any specific solutions into tauri now with the generic solution (not sure). What i would love though is to collect different examples for the generic command (from the aforementioned #9865) in the docs.

Thats okay, but signing with the Trusted Signing requires the Azure CLI, a specific dlib, a manifest.json, so its not that simple. Also I think this will be the recommended (or required?) way to sign apps for Windows (accroding to Microsoft) when it leave the private beta. So I think Tauri should have a built-in way to sign with Trusted Signing.

**Comment by anatawa12** [comment]
Backport is #9902

**Comment by tobyspark** [comment]
> I don't think we'll add any specific solutions into tauri now with the generic solution (not sure). What i would love though is to collect different examples for the generic command (from the aforementioned https://github.com/tauri-apps/tauri/pull/9865) in the docs.

Could I nudge your view a little? As a newcomer to Tauri, it strikes me as odd to be able to supply “platform native” signing credentials for one platform but not another. I get that Azure Trusted Signing is new, and the established-three-year issue certainly sucks, but it’s clearly the direction of travel. 

Also as a newcomer, I assumed without direct support I could still e.g. add the [MS supplied Github action](https://github.com/azure/trusted-signing-action) to my publish workflow as a post-build step. But with a little understanding of what’s happening, of course it needs to happen before the bundling steps in the build process. The reasons why are good, I’m writing this to say there isn’t an escape hatch here. If the custom sign-tool of #9865 allows the provider-specific cloud based flows now pretty much required by EV certificate signing, that’s great, but as @Levminer notes is not adequate for Azure Trusted Signing.

**Comment by tobyspark** [comment]
FWIW, here is the implementation I arrived for Azure Trusted Signing in Tauri on a GitHub actions workflow.

I add a prior step to the Tauri action that installs @Levminer’s CLI tool, and assembles and inserts the `beforeBundleCommand` entry to invoke it into `tauri.conf.json`. You’ll need to know the exact path of the file(s) you want to sign. You then supply the remaining Azure credentials alongside the Apple ones in the Tauri action.

```
      - name: install dependencies (windows)
        if: matrix.platform == 'windows-latest'
        # The approach here uses the cli tool workaround from the GH issue linked below, with some tauri.conf.json manipulation to add the sign command only for windows and pre-populated with secrets
        # https://github.com/tauri-apps/tauri/issues/9578
        # Note Github hosted runners appear to have `az` pre-installed, which is a requirement of trusted-signing-cli
        shell: bash
        env:
          BEFORE_BUNDLE_COMMAND: trusted-signing-cli -e "${{ secrets.AZURE_ENDPOINT }}" -a "${{ secrets.AZURE_CODE_SIGNING_NAME }}" -c "${{ secrets.AZURE_CERT_PROFILE_NAME }}" "${{ github.workspace }}/src-tauri/target/release/deps/<your app>.exe"
        run: |
          cd "$GITHUB_WORKSPACE"
          cat './src-tauri/tauri.conf.json' | jq '.build += {
              "beforeBundleCommand": env.BEFORE_BUNDLE_COMMAND
            }' > './src-tauri/temp.json' && mv ‘./src-tauri/temp.json' './src-tauri/tauri.conf.json'
          cargo install trusted-signing-cli

      - uses: tauri-apps/tauri-action@v0
        env:
          GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
          APPLE_CERTIFICATE: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_CERTIFICATE }}
          APPLE_CERTIFICATE_PASSWORD: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_CERTIFICATE_PASSWORD }}
          APPLE_SIGNING_IDENTITY: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_SIGNING_IDENTITY }}
          APPLE_ID: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_ID }}
          APPLE_PASSWORD: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_PASSWORD }}
          APPLE_TEAM_ID: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_TEAM_ID }}
          AZURE_TENANT_ID: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_TENANT_ID }}
          AZURE_CLIENT_ID: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_CLIENT_ID }}
          AZURE_CLIENT_SECRET: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_CLIENT_SECRET }}
      [...]
```

Now, have I actually got a signed build? No, but the problem seems to be with the credentials I’ve been supplied with, as the Azure Trusted Signing action fails in the same way.

**Comment by tobyspark** [comment]
> At least once MS lifts the stupid 3-years tax history requirement, this should have explicit support in tauri imo (so that it's a direct config option that doesn't require the use of the generic option)

Looks like there’s movement here, this is from a week ago (June 2024): "we are on the process of rolling out this feature for individual maintainers and or young corporations, our expectation is that it should be publicly available in the next upcoming weeks.” – https://github.com/Azure/trusted-signing-action/issues/10#issuecomment-2165678856

**Comment by bra1n** [comment]
> Looks like there’s movement here, this is from a week ago (June 2024): "we are on the process of rolling out this feature for individual maintainers and or young corporations, our expectation is that it should be publicly available in the next upcoming weeks.” – [Azure/trusted-signing-action#10 (comment)](https://github.com/Azure/trusted-signing-action/issues/10#issuecomment-2165678856)

In light of this, can we re-open this ticket to collect all relevant discussions in one issue rather than across multiple new ones, please? @FabianLars @lucasfernog 

Edit: for context, I'm in a similar position as OP in that I would like to have a simple way to use the (official) MS / Azure code sign pipeline with Tauri, one that doesn't need extra steps or external tooling.

**Comment by Levminer** [comment]
> FWIW, here is the implementation I arrived for Azure Trusted Signing in Tauri on a GitHub actions workflow.
> 
> I add a prior step to the Tauri action that installs @Levminer’s CLI tool, and assembles and inserts the `beforeBundleCommand` entry to invoke it into `tauri.conf.json`. You’ll need to know the exact path of the file(s) you want to sign. You then supply the remaining Azure credentials alongside the Apple ones in the Tauri action.
> 
> ```
>       - name: install dependencies (windows)
>         if: matrix.platform == 'windows-latest'
>         # The approach here uses the cli tool workaround from the GH issue linked below, with some tauri.conf.json manipulation to add the sign command only for windows and pre-populated with secrets
>         # https://github.com/tauri-apps/tauri/issues/9578
>         # Note Github hosted runners appear to have `az` pre-installed, which is a requirement of trusted-signing-cli
>         shell: bash
>         env:
>           BEFORE_BUNDLE_COMMAND: trusted-signing-cli -e "${{ secrets.AZURE_ENDPOINT }}" -a "${{ secrets.AZURE_CODE_SIGNING_NAME }}" -c "${{ secrets.AZURE_CERT_PROFILE_NAME }}" "${{ github.workspace }}/src-tauri/target/release/deps/<your app>.exe"
>         run: |
>           cd "$GITHUB_WORKSPACE"
>           cat './src-tauri/tauri.conf.json' | jq '.build += {
>               "beforeBundleCommand": env.BEFORE_BUNDLE_COMMAND
>             }' > './src-tauri/temp.json' && mv ‘./src-tauri/temp.json' './src-tauri/tauri.conf.json'
>           cargo install trusted-signing-cli
> 
>       - uses: tauri-apps/tauri-action@v0
>         env:
>           GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
>           APPLE_CERTIFICATE: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_CERTIFICATE }}
>           APPLE_CERTIFICATE_PASSWORD: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_CERTIFICATE_PASSWORD }}
>           APPLE_SIGNING_IDENTITY: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_SIGNING_IDENTITY }}
>           APPLE_ID: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_ID }}
>           APPLE_PASSWORD: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_PASSWORD }}
>           APPLE_TEAM_ID: ${{ matrix.platform == 'macos-latest' && secrets.APPLE_TEAM_ID }}
>           AZURE_TENANT_ID: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_TENANT_ID }}
>           AZURE_CLIENT_ID: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_CLIENT_ID }}
>           AZURE_CLIENT_SECRET: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_CLIENT_SECRET }}
>       [...]
> ```
> 
> Now, have I actually got a signed build? No, but the problem seems to be with the credentials I’ve been supplied with, as the Azure Trusted Signing action fails in the same way.

Did you set up the permissions? (https://learn.microsoft.com/en-us/azure/trusted-signing/tutorial-assign-roles) If you can't sign with the official action this should be the problem.

**Comment by tobyspark** [comment]
> Did you set up the permissions? (https://learn.microsoft.com/en-us/azure/trusted-signing/tutorial-assign-roles) If you can't sign with the official action this should be the problem.

For the work I’ve got I’m being supplied with them rather than doing it myself. But I’ve read every damn explainer and screenshared with the project owner while they go through it... and yeah, we’ve dotted all the `i`s and crossed all the `t`s as far as we can see.

**Comment by bra1n** [comment]
@tobyspark In case you haven't seen it, this guide has been extremely helpful with getting the cert up and running: https://melatonin.dev/blog/code-signing-on-windows-with-azure-trusted-signing/ (I'm currently still in the validation step unfortunately)

**Comment by tobyspark** [comment]
With good credentials, what [I posted above](https://github.com/tauri-apps/tauri/issues/9578#issuecomment-2178282244) works. But – that’s only signing the app executable, and it is now clear to me that the installer needs to be signed. Does anybody have a working solution for that? e.g.
- there is no `afterBundleCommand` equivalent to `beforeBundleCommand` I could use to invoke the signing within the Tauri build action
- the Wix documentation mentions [signing](https://wixtoolset.org/docs/tools/signing/#signing-windows-installer-packages), but in the context of the `.wixproj` whatever that is. Empirically, adding the `SignMSI` part as a Wix fragment in the Tauri conf doesn’t work. I’m assuming this is trivial to someone who is used to Windows and/or Tauri itself.

Failing the above, I started to just brute-force it with a subsequent Github Action step, but even that is troublesome as the installer filename includes the dynamic version number.

Edit to add – I also tried the back-port of #9865 but it is not part of a v1.x release yet, and I couldn’t figure out the correct invocation of `npm install` to get the v1 branch. `https://gitpkg.now.sh/tauri-apps/tauri/tooling/cli/node?1.x` appears to work, correctly installs as `@tauri-apps/cli` but then on use fails on a platform specific version.

**Comment by FabianLars** [comment]
> Edit to add – I also tried the back-port of https://github.com/tauri-apps/tauri/pull/9865 but it is not part of a v1.x release yet, and I couldn’t figure out the correct invocation of npm install to get the v1 branchEdit to add – I also tried the back-port of https://github.com/tauri-apps/tauri/pull/9865 but it is not part of a v1.x release yet, and I couldn’t figure out the correct invocation of npm install to get the v1 branch

You can't really do that, only the cargo based cli works from git (because the npm based cli is 2 npm packages and you only get the wrapper one from git). You can however just install the cli in CI with a new step that runs `cargo install tauri-cli --git https://github.com/tauri-apps/tauri --branch 1.x` and then add `tauriScript: cargo tauri` to the tauri-action step. (Works locally as well of course)

**Comment by Levminer** [comment]
Here's my action: https://github.com/Levminer/authme/blob/dev/.github/workflows/release-artifacts.yml
And here is a helper script I'm using: https://github.com/Levminer/authme/blob/dev/scripts/sign.js

But basically:

1. Sign the app in the beforeBundleCommand
2. I get the version from the package.json
3. Using that version I sign the msi or nsis

**Comment by bra1n** [comment]
> Here's my action: https://github.com/Levminer/authme/blob/dev/.github/workflows/release-artifacts.yml And here is a helper script I'm using: https://github.com/Levminer/authme/blob/dev/scripts/sign.js
> 
> But basically:
> 
> 1. Sign the app in the beforeBundleCommand
> 2. I get the version from the package.json
> 3. Using that version I sign the msi or nsis

I looked into that - thanks for sharing! However, it's not using the Tauri Github Action, rather it does all the steps manually, right? I would like to use the Github Action because it does all the individual steps, including building, bundling, update sig generation and uploading in one place... :thinking:

**Comment by bra1n** [comment]
> With good credentials, what [I posted above](https://github.com/tauri-apps/tauri/issues/9578#issuecomment-2178282244) works. But – that’s only signing the app executable, and it is now clear to me that the installer needs to be signed. Does anybody have a working solution for that?

I finally did get it working! However, it requires Tauri v2, as it uses the new config parameter from #9865.

Here's the relevant Github Action block, taken from @Levminer's approach:
```yaml
      - name: install dependencies (windows only)
        if: matrix.platform == 'windows-latest'
        shell: bash
        env:
          # replace Azure credentials before running
          WINDOWS_SIGN_COMMAND: trusted-signing-cli -e https://weu.codesigning.azure.net/ -a [AZURE_CODE_SIGNING_NAME] -c [AZURE_CERT_PROFILE_NAME] %1
        run: |
          cd "$GITHUB_WORKSPACE"
          cat './src-tauri/tauri.conf.json' | jq '.bundle .windows += {"signCommand": env.WINDOWS_SIGN_COMMAND}' > './src-tauri/temp.json' && mv './src-tauri/temp.json' './src-tauri/tauri.conf.json'
          cargo install trusted-signing-cli
``` 
I hard-coded the non-secret AZURE credentials there, but you can also use env vars if you want. The `signCommand` signs the built executable before bundling and the bundled installers afterwards. Since its run directly from within the Tauri CLI, you can then use the normal Tauri Github Action to take care of the Updater signing and uploading:
```yaml
      - uses: tauri-apps/tauri-action@v0
        env:
          GITHUB_TOKEN: ${{ secrets.GH_TOKEN }}
          TAURI_SIGNING_PRIVATE_KEY: ${{ secrets.TAURI_SIGNING_PRIVATE_KEY }}
          APPLE_CERTIFICATE: ${{ secrets.APPLE_CERTIFICATE }}
          APPLE_CERTIFICATE_PASSWORD: ${{ secrets.APPLE_CERTIFICATE_PASSWORD }}
          APPLE_SIGNING_IDENTITY: ${{ secrets.APPLE_SIGNING_IDENTITY }}
          APPLE_ID: ${{ secrets.APPLE_ID }}
          APPLE_PASSWORD: ${{ secrets.APPLE_PASSWORD }}
          APPLE_TEAM_ID: ${{ secrets.APPLE_TEAM_ID }}
          AZURE_TENANT_ID: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_TENANT_ID }}
          AZURE_CLIENT_ID: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_CLIENT_ID }}
          AZURE_CLIENT_SECRET: ${{ matrix.platform == 'windows-latest' && secrets.AZURE_CLIENT_SECRET }}
        with:
          updaterJsonPreferNsis: true
          tagName: v__VERSION__ # the action automatically replaces \_\_VERSION\_\_ with the app version.
          releaseName: 'v__VERSION__ ${{ env.RELEASE_TITLE }}'
          releaseBody: "${{ env.RELEASE_NOTES }}"
          releaseDraft: true
          prerelease: false
          args: ${{ matrix.args }}
```
With this workflow I was able to get signed installers and also the installed app itself will run without any warning messages.

Edit: updated the config to remove the need for `beforeBundleCommand`, as pointed out by @FabianLars below.

**Comment by FabianLars** [comment]
If you use the signCommand config you can remove beforeBundleCommand. Tauri will use the signCommand for everything that needs to be signed, including the inner executable.

**Comment by bra1n** [comment]
> If you use the signCommand config you can remove beforeBundleCommand. Tauri will use the signCommand for everything that needs to be signed, including the inner executable.

Oh nice, I didn't know that! I'll adjust the above code then.

**Comment by jacobgorm** [comment]
One caveat with all the approaches above is that they will expose your Azure secrets to rogue code in any build.rs script you or your dependencies happen to drag via crates.io. I believe that tauri2 supports doing the bundling separately from the build, which would increase security.

**Comment by tobyspark** [comment]
Thanks all. In hindsight, why did I spent time trying to wrangle the back-port, when the upgrade to Tauri v2 proved quick and painless =]

**Comment by bra1n** [comment]
> One caveat with all the approaches above is that they will expose your Azure secrets to rogue code in any build.rs script you or your dependencies happen to drag via crates.io. I believe that tauri2 supports doing the bundling separately from the build, which would increase security.

Good point! Do you happen to know how one could get an auxiliary build tool integrated into the Tauri build chain the easiest? Just run `cargo add trusted-signing-cli` in `/src-tauri` and commit the updated files?

**Comment by corentin35000** [comment]
I post here, seen with FabianLars already, when building tauri via: tauri build -- --target x86_64-pc-windows-msvc (npm), signing via signCommand fails because it tries to sign the .exe located in: /target/release/.. instead of /target/x86_64-pc-windows../, so this fails. But it works well directly with "tauri build" (without the --target), for those who wish to compile for several architectures, to get around this, you must change the compilation chain via: rustup default stable-i686-pc-windows- msvc (for x86 for example) and be able to use tauri build and the signCommand which will work. I hope for a fix from Tauri. For those who have this problem.

**Comment by leet4tari** [comment]
Based on @bra1n post, but staying with ```Tauri v1```, kept our step a little cleaner and easier to read, I think:
```yaml
      - name: Azure code-signing setup (windows only)
        if: ${{ ( startsWith(runner.os,'Windows') ) && ( env.AZURE_TENANT_ID != '' ) }}
        shell: bash
        env:
          AZURE_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
          WINDOWS_SIGN_COMMAND: trusted-signing-cli -e https://weu.codesigning.azure.net/ -a [AZURE_CODE_SIGNING_NAME] -c [AZURE_CERT_PROFILE_NAME] %1
        run: |
          cd "$GITHUB_WORKSPACE/src-tauri"
          echo "{}" | \
            jq '.tauri .bundle .windows += {"signCommand": "${{ env.WINDOWS_SIGN_COMMAND }}"}' > \
              ./tauri.windows.conf.json
          cat ./tauri.windows.conf.json
          cargo install trusted-signing-cli
```
The yaml block only runs if there are the Azure secrets, so forked repos should still be able to use the workflow. Plus only windows targets include the ```tauri.windows.conf.json``` file.  
  
Hope this helps somebody else.

**Comment by don41382** [comment]
It's not absolute related, but It took my almost a half day to figure out, why the signing tool didn't work. I always got:

> SignTool Error: No certificates were found that met all the given criteria.

I had to switch in https://portal.azure.com/#view/Microsoft_AAD_RegisteredApps/ApplicationsListBlade the `App Registration` -> `<service-principal>` -> `Authentication` and switch it to allow public client flow.

<img width="699" alt="image" src="https://github.com/user-attachments/assets/c7488bff-4f74-4be2-84fc-971da090a761">

And also: Don't forget to add the roles `Trusted Signing Certificate Profile Signer` and `Trusted Signing Identity Verifier` to your service-principal in your `Trusted Signing Account`.

Maybe this helps someone and save time or future me, who will run into the same trap.

**Comment by apicize** [comment]
Hi, I'm working my way through trying to get this set up.  I have successfully created a Trusted Signing Account as an individual developer, including the steps of Identity validation and Certificate profile.

This is my first experience using Trusted Signing, and .MSI signing in general, so apologies if the questions are dumb, answers are obvious etc.  If there is a page that explains this, happy to read that as well.  I tried [this](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/walkthrough-register-app-azure-active-directory) document referenced in the [trusted-signing-cli](https://github.com/Levminer/trusted-signing-cli) project, and created an App Registration, but in attempting to add the referenced API Permission, "Dataverse" doesn't come up in Request API Permissions referenced therein.  

Basically, I'm trying to figure out where to get these values referenced [above](https://github.com/tauri-apps/tauri/issues/9578#issuecomment-191411269):

* AZURE_CODE_SIGNING_NAME:  App registration name? Trusted signing acct name?  Something else?
* AZURE_CERT_PROFILE_NAME:  Certificate profile name set up in Trusted Signing Account?
* AZURE_TENANT_ID: (got this)
* AZURE_CLIENT_ID: ??? Some OAuth-ish credential created somewhere?
* AZURE_CLIENT_SECRET: ??? Some OAuth-ish credential created somewhere?

I think I got the pieces created, just unsure how to knit them together.

Thanks!

**Comment by Levminer** [comment]
> Hi, I'm working my way through trying to get this set up. I have successfully created a Trusted Signing Account as an individual developer, including the steps of Identity validation and Certificate profile.
> 
> This is my first experience using Trusted Signing, and .MSI signing in general, so apologies if the questions are dumb, answers are obvious etc. If there is a page that explains this, happy to read that as well. I tried [this](https://learn.microsoft.com/en-us/power-apps/developer/data-platform/walkthrough-register-app-azure-active-directory) document referenced in the [trusted-signing-cli](https://github.com/Levminer/trusted-signing-cli) project, and created an App Registration, but in attempting to add the referenced API Permission, "Dataverse" doesn't come up in Request API Permissions referenced therein.
> 
> Basically, I'm trying to figure out where to get these values referenced [above](https://github.com/tauri-apps/tauri/issues/9578#issuecomment-191411269):
> 
> * AZURE_CODE_SIGNING_NAME:  App registration name? Trusted signing acct name?  Something else?
> * AZURE_CERT_PROFILE_NAME:  Certificate profile name set up in Trusted Signing Account?
> * AZURE_TENANT_ID: (got this)
> * AZURE_CLIENT_ID: ??? Some OAuth-ish credential created somewhere?
> * AZURE_CLIENT_SECRET: ??? Some OAuth-ish credential created somewhere?
> 
> I think I got the pieces created, just unsure how to knit them together.
> 
> Thanks!

Those are all valid questions, its kind of complicated thanks to Microsoft.

AZURE_CODE_SIGNING_NAME:  Trusted signing account name
AZURE_CERT_PROFILE_NAME: Yes, the name of the certificate inside your trusted signing account
AZURE_CLIENT_ID: You get this value when you create an app registration
AZURE_CLIENT_SECRET: You also get this when creating an app registration

I will make a PR adding docs to the tauri docs website.

EDIT: Also read this its a very good guide: https://melatonin.dev/blog/code-signing-on-windows-with-azure-trusted-signing/

**Comment by Levminer** [comment]
Docs are available here: https://tauri.app/distribute/sign/windows/#azure-code-signing

**Comment by tobyspark** [comment]
Props on the documentation. I’ve just made a `singing` → `signing` typo [PR](https://github.com/tauri-apps/tauri-docs/pull/2964)

**Comment by apicize** [comment]
After the help on this thread and a bunch of trial and error, I got the GitLab action working.  One thing that made it a little tough for me using the trusted-signing-cli crate was that I did not get a lot of feedback when stuff didn't work.  I ended up having to manually go through the steps in Powershell to figure out what was going wrong (Azure configuration is a PITA, and I'll work on some supplementary documentation that I'll submit to the Tauri project after catching up with some stuff neglected around the house).

Below is an alternate GitHub action using Powershell that implements Azure Trusted Signing without any crate dependencies.  It has to be placed before the build step.  It does the work of looking for signtool.exe, grabbing Download Microsoft.Trusted.Signing.Client, writing the Authenticode digest JSON, and updateing tauri.conf.json.   If somebody is having trouble getting trusted-signing-cli to work, this method is more verbose but may provide helpful error info.

Environment variable values used in the step are described as follows.  In the step shown, all the values are stored as action vars, except for AZURE_TENANT_ID and AZURE_CLIENT_SECRET which are stored as action secrets

* **AZURE_TENANT_ID**: required, found in App Registration Overview as "Directory (tenant) ID"
* **AZURE_CLIENT_ID**:  required, found in App Registration Overview as "Application (client) ID"
* **AZURE_CLIENT_SECRET**: required, created in App Registartion
* **AZURE_ENDPOINT**: required, found in Trusted Signing Account as "Account URI"
* **AZURE_CERT_PROFILE_NAME**: required, found in Trusted Signing Account, Objects, Certificate profiles
* **AZURE_CODE_SIGNING_NAME**: required, found in Trusted Signing Account as Name
* **TAURI_CONF_JSON**: optional, relative location of tauri.conf.json, defaults to `./src-tauri/tauri.conf.json`

```yaml
    - name: Add Windows Trusted Signing
      shell: pwsh
      if: matrix.platform == 'windows-latest'
      env:
        AZURE_TENANT_ID: ${{ secrets.AZURE_TENANT_ID }}
        AZURE_CLIENT_ID: ${{ vars.AZURE_CLIENT_ID }}
        AZURE_CLIENT_SECRET: ${{ secrets.AZURE_CLIENT_SECRET }}
        AZURE_ENDPOINT: ${{ vars.AZURE_ENDPOINT }}
        AZURE_CERT_PROFILE_NAME: ${{ vars.AZURE_CERT_PROFILE_NAME }}
        AZURE_CODE_SIGNING_NAME: ${{ vars.AZURE_CODE_SIGNING_NAME }}
        TAURI_CONF_JSON: "./app/src-tauri/tauri.conf.json"
      run: |
        $tempFolderPath = Join-Path $env:Temp "mtsc"
        $mtscName = Join-Path $tempFolderPath "mtsc.nuget.zip"
        $dlibName = Join-Path $tempFolderPath "bin" | Join-Path -ChildPath "x64" | Join-Path -ChildPath "Azure.CodeSigning.Dlib.dll" 
        $dmdName = Join-Path $tempFolderPath "metadata.json"
        $tauriConfJson = $env:TAURI_CONF_JSON ?? "./src-tauri/tauri.conf.json"

        if (-not(Test-Path -path $tauriConfJson)) {
          Throw "Unable to locate Tauri configuration at $taurConfJson"
        }

        $signTool = Get-Command "signtool.exe" -ErrorAction SilentlyContinue
        if ($signTool) {
          $signToolPath = Split-Path $signTool
        } else {
          if (Test-Path -path "C:\Program Files (x86)\Windows Kits\10\bin\10.0.22000.0\x64\signtool.exe") {
            $signToolpath = "C:\Program Files (x86)\Windows Kits\10\bin\10.0.22000.0\x64"
          } else {
            Write-Host "Searching for signtool.exe in C:\Program Files (x86)"
            $signtoolMatches = Get-Childitem –Path "C:\Program Files (x86)" -Include signtool.exe -File -Recurse -ErrorAction SilentlyContinue
            if ($signtoolMatches.Length -gt 0) {
              $signToolPath = Split-path $signtoolMatches[0].FullName
            } else {
              Write-Host "Searching for signtool.exe in C:\Program Files"
              $signtoolMatches = Get-Childitem –Path "C:\Program Files" -Include signtool.exe -File -Recurse -ErrorAction SilentlyContinue
              if ($signtoolMatches.Length -gt 0) {
                $signToolPath = Split-path $signtoolMatches[0].FullName
              }
            }
          }
        }

        if ($signToolPath) {
          Write-Host "Appending location of signtool.exe to PATH: $signToolPath"
          Add-Content $env:GITHUB_PATH $signToolPath
        } else {
          Throw "Unable to locate signtool.exe"
        }

        if (-not(Test-Path -path $tempFolderPath)) {
            New-Item -Type Directory -Path $tempFolderPath | Out-Null
            Write-Host "Download Microsoft.Trusted.Signing.Client to $mtscName"
            $global:ProgressPreference = "SilentlyContinue"
            Invoke-WebRequest "https://www.nuget.org/api/v2/package/Microsoft.Trusted.Signing.Client/" -OutFile $mtscName
            Write-Host "Extracting $mtscName"
            Expand-Archive $mtscName -Force -DestinationPath $tempFolderPath
            $global:ProgressPreference = 'Continue'
        }

        Write-Host "Saving Autheticode digest to $dmdName"
        $metadata = [PSCustomObject]@{
            Endpoint               = $env:AZURE_ENDPOINT
            CodeSigningAccountName = $env:AZURE_CODE_SIGNING_NAME
            CertificateProfileName = $env:AZURE_CERT_PROFILE_NAME
        }
        $metadataAsJson = ConvertTo-Json -Depth 100 $metadata
        Set-Content -Path $dmdName -Value $metadataAsJson

        $batchName = Join-Path $tempFolderPath "signtool.bat"
        Set-Content -Path $batchName -Value """$signToolPath"" sign /v /fd SHA256 /tr http://timestamp.acs.microsoft.com /td SHA256 /dlib $dlibName /dmdf $dmdName %1"

        $signCommand = "signtool sign /v /fd SHA256 /tr http://timestamp.acs.microsoft.com /td SHA256 /dlib $dlibName /dmdf $dmdName %1"
        Write-Host "Updating Windows signCommand in $tauriConfJson to $signCommand"
        $tauriConfigData = ConvertFrom-Json(Get-Content $tauriConfJson -Raw)
        $tauriConfigData.bundle.windows | Add-Member -Force -MemberType NoteProperty -Name signCommand -Value $signCommand
        $tauriConfigDataAsJson = ConvertTo-Json -Depth 100 $tauriConfigData
        Set-Content -Path $tauriConfJson -Value $tauriConfigDataAsJson

        Write-Host "Login to Azure"
        az login --service-principal --output none -t $env:AZURE_TENANT_ID -u $env:AZURE_CLIENT_ID -p $env:AZURE_CLIENT_SECRET
```

**Comment by dceddia** [comment]
I got this working thanks to the official Tauri docs and the `trusted-signing-cli` tool which was great to find (after trying and failing to cobble together the signtool command myself).

I hit one weird thing that maybe the docs could mention and provide in the default `trusted-signing-cli` example: the `-d` or `--description` flag is important if you want the name shown by the MSI installer to match your product name.

The trusted-signing-cli's help explains this flag but I missed it on the first pass. Without `-d YOUR_APP_NAME` it will quite literally be a random string of characters (randomly re-generated every time you run the MSI, even!)

**Comment by leet4tari** [comment]
You need ```trusted-signing-cli``` v0.5.0 for the description arg to work and it only takes one none-spaced nor none-special character word. Started looking at if it was a limitation of ```trusted-signing-cli``` or how it processes the command line args. Tried with single quotes, double quotes and ended up using camelCase, until we have more time to test further. Using ```azure/trusted-signing-action```  and ```description``` works with multiple words for the description.

**Comment by Levminer** [comment]
> You need `trusted-signing-cli` v0.5.0 for the description arg to work and it only takes one none-spaced nor none-special character word. Started looking at if it was a limitation of `trusted-signing-cli` or how it processes the command line args. Tried with single quotes, double quotes and ended up using camelCase, until we have more time to test further. Using `azure/trusted-signing-action` and `description` works with multiple words for the description.

Can you post the command you tried to use? And which shell did you try to use? I tried using bash,powershell,zsh and I couldn't reproduce this. Please file and issue: https://github.com/Levminer/trusted-signing-cli

**Comment by dceddia** [comment]
@Levminer your reply reminded me of an issue I ran into while I was setting everything up.

My tauri signCommand looked like this, where I had single quotes around the -a parameter:

```
"signCommand": "trusted-signing-cli -e https://wus.codesigning.azure.net -a 'code-signing-for-app' -c Param -d Param %1",
```

The command it tried to run had escaped the single quotes somehow. I'm not 100% certain now, but I think they were turned into percent encoding as %27 or something. This struck me as something Tauri was probably doing when it formed the command more than something trusted-signing-cli was doing, and anyway my string didn't require the quotes so I took them out.

**Comment by Levminer** [comment]
Yes you are correct, this is a limitation of the signCommand: https://tauri.app/reference/config/#customsigncommandconfig

Also, filed a PR for the description: https://github.com/tauri-apps/tauri-docs/pull/3193

**Comment by space-man-rob** [comment]
Is there a way to combine azure signing with the updater plugin?  It breaks the signature flow

**Comment by FabianLars** [comment]
If you use the sign command approach mentioned above it should work with the updater as well.
If you can't use that (please tell us why) you can always use the `tauri signer sign` command to generate the updater artifacts afterwards.

**Comment by bra1n** [comment]
> Is there a way to combine azure signing with the updater plugin? It breaks the signature flow

I can confirm that it does work out of the box with the signCommand config and the following addition to the Github Actions config:
```yaml
      - name: install dependencies (windows only)
        if: matrix.platform == 'windows-latest'
        run: cargo install trusted-signing-cli@0.5.0
```
The updater plugin integrates seamlessly into this, assuming you're on Tauri v2.

**Comment by space-man-rob** [comment]
@FabianLars @bra1n thanks, got it working

**Comment by FTCHD** [comment]
You can check the [`rclone`](https://github.com/rclone-ui/rclone-ui) repo the see how it's implemented **[here](https://github.com/rclone-ui/rclone-ui/blob/main/.github/workflows/release-windows.yml)**

Really simple and works with the updater plugin!
